### ISSUE TITLE
Error requesting miele data: Cannot connect to host api.mcs3.miele.com:443 ssl:default [Name has no usable address]

### ISSUE THREAD

**Issue by djwafelap** [astrandb/miele#471]
### The problem

The log reports the following error while the Miele integration is "not available":

Error requesting miele data: Cannot connect to host api.mcs3.miele.com:443 ssl:default [Name has no usable address]

It happens so often that the integration is not really reliable and it is hardly possible to record the work of the dishwasher. 

The error looks to me like this is a server problem at Miele or some kind of limitation for third party customers. 

Do you have this so often and for so long? 

Cheers
Rado

### Version of miele integration having the issue?

miele-08.23

### Version of Home Assistant Core having the issue?

core-2024.4.4

**Comment by astrandb** [comment]
Please make sure that you are using this integration. The version `miele-08.23` is not from this integration.

**Comment by astrandb** [comment]
Thanks, then I know that you are on the right field...
This is not a normal behaviour of the integration. Please attach relevant lines from the full HA log.
BTW Are you using the AdGuard add-on. I have got a few reports in another context that it causes this type of interference.

**Comment by djwafelap** [comment]
Hi,
I don't use AdGuard.
Here are all entries regarding miele in the full HA log:

"2024-04-28 18:42:43.896 ERROR (MainThread) [custom_components.miele] Error requesting miele data: Cannot connect to host api.mcs3.miele.com:443 ssl:default [Name has no usable address]
2024-04-28 18:42:43.896 DEBUG (MainThread) [custom_components.miele] Finished fetching miele data in 0.079 seconds (success: False)
2024-04-28 18:42:43.900 DEBUG (MainThread) [custom_components.miele] Finished fetching miele data in 0.004 seconds (success: False)
"

And here is the full entry of the error in the system log:
"Dieser Fehler wurde von einer benutzerdefinierten Integration verursacht

Logger: custom_components.miele
Quelle: helpers/update_coordinator.py:328
Integration: Miele (Dokumentation, Probleme)
Erstmals aufgetreten: 27. April 2024 um 02:24:45 (123 Vorkommnisse)
Zuletzt protokolliert: 18:46:40

Error requesting miele data: Cannot connect to host api.mcs3.miele.com:443 ssl:default [Name has no usable address]
Error requesting miele data: Cannot connect to host api.mcs3.miele.com:443 ssl:default [Try again]"


I also had this error before the integration happened to be running again:
"General error while obtaining the access token."
![image](https://github.com/astrandb/miele/assets/125517688/8f58e703-06f2-4f3b-b7e7-ca83409b6a8b)

Could it perhaps have something to do with that? 




> Thanks, then I know that you are on the right field... This is not a normal behaviour of the integration. Please attach relevant lines from the full HA log. BTW Are you using the AdGuard add-on. I have got a few reports in another context that it causes this type of interference.

**Comment by astrandb** [comment]
`Allgemeiner Fehler während der Einholung des Zugriffstokens.` appear if there is an authentication error when setting up the integration the first time. The result is unpredictable if you manage to pass this error message.
I suggest that you uninstall the integration while making sure that the oauth credentials are removed also. Then install the integration again.

**Comment by djwafelap** [comment]
hm, after complete remove and reinstalling, the problem is still the same :(
is it possible that miele somehow blocked me or put me on a kind of black list due to too many requests???

**Comment by astrandb** [comment]
I don't think that Miele is blocking you, There should be "429"-errors in the log then. It must something in your setup, The integration works for more than 6600 users.

You can verify that your app credentials work on https://www.miele.com/developer/swagger-ui/index.html - Click `Authorize`

**Comment by djwafelap** [comment]
Is there any setup I can do???



> I don't think that Miele is blocking you, There should be "429"-errors in the log then. It must something in your setup, The integration works for more than 6600 users.
> 
> You can verify that your app credentials work on https://www.miele.com/developer/swagger-ui/index.html - Click `Authorize`

**Comment by astrandb** [comment]
Check that you can authorize on Miele developer site, link above. Other users have noted problems if their password includes any special characters.

**Comment by djwafelap** [comment]
looks like i am logged in (???)
![image](https://github.com/astrandb/miele/assets/125517688/2c80b9cb-1505-46a4-a164-637584f39548)



> Check that you can authorize on Miele developer site, link above. Other users have noted problems if their password includes any special characters.

**Comment by djwafelap** [comment]
Does the "Application Name" have a special purpose?

I have also noticed that the integration hangs on the first attempt at a new installation until I cancel it. No files have been copied to the /config/custom_components/miele directory. The integration is only installed on the second attempt. 

Could this have something to do with it?

**Comment by astrandb** [comment]
> Does the "Application Name" have a special purpose?

In what context? If when entering application credentials it is just a descriptive name. In HA you can store multiple sets for an integration. You will then be prompted for which credentials set you want to use. Mainly usful for developers that want to swap between different Miele accounts.

I think we need to be very specific on the detailss here: 
1. You typically use HACS or a manual copy to DOWNLOAD the integration to /config/custom_components/
1. Restart HA to let it learn about the downloaded integration.
1. INSTALL the integration from Settings->Integrations by selecting Miele integration. Many times the Miele integration will autodiscover any Miele appliances present on your LAN and prompt you to install the integration
1. If it is a first time install you will be prompted to enter application credentials.
1. You will be redirected to the Miele site to verify your account.
1. You should the be directed back to your HA and the setup will complete.

At what point are you getting errors?

> No files have been copied to the /config/custom_components/miele directory.

Sounds very odd.

Can you please tell about your installation:
Environment: HassOS, VM, docker...
HA core version:
HACS version - if any:

**Comment by djwafelap** [comment]
ok, just thought the application name could have a special meaning. if it's only a description, it should not make problems.


1. yes, i use hacs for downloading the integrations
![image](https://github.com/astrandb/miele/assets/125517688/17be673f-4b4a-4b90-83c9-ccf282048e58)


here is what happens next: it looks like miele is beeing installed, but nothing happens except of the rotating circle. 
![image](https://github.com/astrandb/miele/assets/125517688/9ff65db6-f6f7-4220-a9e5-b8b97cb28aa5)
i have to cancel this process and start downloading once again, then it works

2. after downloading, i restart HA

3.  then i install the integration
![image](https://github.com/astrandb/miele/assets/125517688/0b9aaaf3-026a-4089-aa03-8306b42b9301)

i don't know why at this point it knows the model of my dishwasher...
![image](https://github.com/astrandb/miele/assets/125517688/d2bfa611-d1f3-41ec-b3f2-4532131b6fed)

4. entering the credentials
![image](https://github.com/astrandb/miele/assets/125517688/e8f0221f-cdc0-46c3-a3df-05f459e66bcb)


5. verifying my account
![image](https://github.com/astrandb/miele/assets/125517688/c0ae3673-edd0-4498-81d5-d5829e387d74)


![image](https://github.com/astrandb/miele/assets/125517688/aa2cce27-fc61-4fa2-995d-c785d14a8e0f)


![image](https://github.com/astrandb/miele/assets/125517688/b2a443fa-b59a-4a6b-b763-5a1627f7d9e9)


6. back on the integrations screen, i get the mentioned error

![image](https://github.com/astrandb/miele/assets/125517688/b4507fee-ef80-4d74-bedd-90ebf714a375)



----------------------------
my installation:
Environment: HassOS 12.2, installed on an intel NUC
HA core version: 2024.5.0
HACS version: 1.34.0

**Comment by astrandb** [comment]
It is the redirect back to HA that fails. Please make sure that you can access your HA instance with the marked url. Otherwise you should edit the url by clicking the pencil and change it to the url you are currently using.
![image](https://github.com/astrandb/miele/assets/5016010/0d56deb0-09c0-4d88-87d0-afbc30e6d2a2)

**Comment by djwafelap** [comment]
Thanks for the hint! It certainly makes sense to place the HA server here and i will try out. But are you sure that this is the only reason? I find it strange that the integration occasionally works after a few dozen attempts and only has frequent dropouts and after a while comes back. Wouldn't it fail permanently if it is the redirect?

![image](https://github.com/astrandb/miele/assets/125517688/49df47ab-204e-4c1b-8a5f-a7682ef59b24)

**Comment by djwafelap** [comment]
I just tried it and I get the same error, both with the local IP address of HA and with the external domain... :(

**Comment by djwafelap** [comment]
no more idea what it could be?
